pkg/reconcilers/base_reconciler.go

```diff
@@ -0,0 +1,137 @@
+package reconcilers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/3scale/3scale-operator/pkg/common"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var log = logf.Log.WithName("reconcilers")
+
+// MutateFn is a function which mutates the existing object into it's desired state.
+type MutateFn func(existing, desired common.KubernetesObject) (bool, error)
+
+func CreateOnlyMutator(existing, desired common.KubernetesObject) (bool, error) {
+	return false, nil
+}
+
+type BaseReconciler struct {
+	client          client.Client
+	scheme          *runtime.Scheme
+	apiClientReader client.Reader
+	ctx             context.Context
```

Comment:
I'm curious about this. What's the usefulness on being able to pass a context.Context object and not keep using the context.TODO directly? Do we have any use case?

Reply:
The context allows some features, like cancelling operations. Thus, the same context instance should be used in all operations. We are not using them, but this is preparation for that. 

It is commonly used pattern in operators.

Reply:
:+1: Do you have any example of an operator making use of it? It seems interesting in that case

Reply:
for instance, Keycloak:
https://github.com/keycloak/keycloak-operator/blob/master/pkg/controller/keycloak/keycloak_controller.go#L53